File: gcpc2024/headlineheat/submissions/accepted/erik.java
import java.io.*;
import java.util.*;

public class erik {
    static final int N = 3 * 100000 + 5;
    static final int C = 27;
    static final int K = 3;
    static final long[] P = {1000000007, 1000000009, 1000000021};
    static final long[] B = {123456789, 987654321, 543212345};

    static int M = 1;
    static int[] lnk = new int[N];
    static int[] val = new int[N];
    static int[] par = new int[N];
    static long[][] hash = new long[N][K];
    static int[][] go = new int[N][C];

    static {
        Arrays.fill(lnk, -1);
        Arrays.fill(par, -1);
    }

    public static int getLink(int k) {
        if (lnk[k] == -1) {
            lnk[k] = (k == 0 || par[k] == 0) ? 0 : walk(getLink(par[k]), val[k]);
        }
        return lnk[k];
    }

    public static int walk(int k, int c) {
        if (go[k][c] == 0) {
            return (k == 0) ? 0 : walk(getLink(k), c);
        }
        return go[k][c];
    }

    public static boolean match(String s) {
        int k = 0;
        long[] res = new long[K];
        for (char c : s.toCharArray()) {
            k = walk(k, (c == ' ') ? 0 : c - 'a' + 1);
            for (int i = 0; i < K; i++) {
                res[i] += hash[k][i] - P[i] * (res[i] + hash[k][i] >= P[i] ? 1 : 0);
            }
        }
        return Arrays.equals(res, new long[K]);
    }

    public static int insert(String s) {
        int u = 0;
        for (char c : s.toCharArray()) {
            int ch = (c == ' ') ? 0 : c - 'a' + 1;
            if (go[u][ch] == 0) {
                go[u][ch] = M;
                par[M] = u;
                val[M] = ch;
                M++;
            }
            u = go[u][ch];
        }
        return u;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        int[] fin = new int[n];
        long[] cur = new long[K];
        Arrays.fill(cur, 1);

        for (int i = 0; i < n; i++) {
            fin[i] = insert(br.readLine().trim());
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = fin[Integer.parseInt(st.nextToken()) - 1];
            int b = fin[Integer.parseInt(st.nextToken()) - 1];
            for (int j = 0; j < K; j++) {
                cur[j] = cur[j] * B[j] % P[j];
                hash[a][j] -= cur[j] - P[j] * (hash[a][j] - cur[j] < 0 ? 1 : 0);
                hash[b][j] += cur[j] - P[j] * (hash[b][j] + cur[j] >= P[j] ? 1 : 0);
            }
        }

        for (int i = 0; i < k; i++) {
            if (match(br.readLine().trim())) {
                out.println("yes");
            } else {
                out.println("no");
            }
        }

        out.close();
    }
}

File: gcpc2024/headlineheat/submissions/accepted/erik.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(2000000)
N, C, K, P, B = 3 * 10**5 + 5, 27, 3, [10**9 + 7, 10**9 + 9, 10**9 + 21], [123456789, 987654321, 543212345]
M, lnk, val, par, hash, go = 1, [-1] * N, [0] * N, [-1] * N, [[0] * K for _ in range(N)], [[0] * C for _ in range(N)]

def get_link(k):
	if lnk[k] == -1:
		lnk[k] = 0 if k == 0 or par[k] == 0 else walk(get_link(par[k]), val[k])
	return lnk[k]

def walk(k, c):
	if go[k][c] == 0:
		return 0 if k == 0 else walk(get_link(k), c)
	return go[k][c]

def match(s):
	k, res = 0, [0] * K
	for c in s:
		k = walk(k, 0 if c == ' ' else ord(c) - ord('a') + 1)
		for i in range(K):
			res[i] += hash[k][i] - P[i] * (res[i] + hash[k][i] >= P[i])
	return res == [0] * K

def insert(s):
	global M
	u = 0
	for c in s:
		c = 0 if c == ' ' else ord(c) - ord('a') + 1
		if go[u][c] == 0:
			go[u][c] = M
			par[M] = u
			val[M] = c
			M += 1
		u = go[u][c]
	return u

n, m, k = map(int, input().split())
fin, cur = [insert(input().strip()) for _ in range(n)], [1] * K
for _ in range(m):
	a, b = map(lambda x: fin[int(x) - 1], input().split())
	for i in range(K):
		cur[i] = cur[i] * B[i] % P[i]
		hash[a][i] -= cur[i] - P[i] * (hash[a][i] - cur[i] < 0)
		hash[b][i] += cur[i] - P[i] * (hash[b][i] + cur[i] >= P[i])
sys.stdout.write("\n".join(["yes" if match(input().strip()) else "no" for _ in range(k)]))